pdd/trainer.py

```diff
@@ -0,0 +1,161 @@
+from .metrics import knn_acc
+from .plot import plot
+import numpy as np
+from scipy.spatial.distance import cosine
+import torch
+from tqdm import tqdm
+
+COUNT_NEIGHBOR_EXP_1 = 1
+COUNT_NEIGHBOR_EXP_2 = 3
+METRIC_KNN = cosine
+COLORS = ['#00ffff', '#000000', '#0000ff', '#ff00ff',
+          '#808080', '#008000', '#00ff00', '#800000',
+          '#000080', '#808000', '#800080', '#ff0000',
+          '#c0c0c0', '#008080', '#ffff00']
+
+
+def forward_inputs_into_model(loader, model, device, batchsize):
+    X = []
+    y = []
+    with torch.no_grad():
+        for batch_idx, (inputs, targets) in enumerate(loader):
+            inputs, targets = inputs.to(device), targets.to(device)
+            outputs = model(inputs).detach().cpu().numpy()
+            targets = targets.detach().cpu().numpy()
+            if (outputs.shape[0] == batchsize):
+                X.append(outputs)
+                y.append(targets)
+    return np.vstack(X), np.hstack(y)
+
+
+def save_model(model, optimizer):
+    torch.save(model, 'triplet_model.pt')
+    torch.save(model.state_dict(), 'triplet_weight.pt')
+    torch.save(optimizer.state_dict, 'optim.pt')
+
+
+class TripletTrainer(object):
+    def __init__(self,
+                 model,
+                 optimizer,
+                 tri_train_load,
+                 epochs,
+                 tri_test_load,
+                 batch_size,
+                 KNN_train_data_load,
+                 KNN_test_data_load,
```

Comment:
- knn_train_loader or train_knn_loader
- knn_test_loader (test_knn_loader)
the same way as for the triplet_loader